packages/dev/serializers/src/USDZ/USDZSerializer.ts

```diff
@@ -0,0 +1,732 @@
+                    float horizontalAperture = ${USDZExport.FilmGauge.toPrecision(USDZExport.Precision)} 
+                    token projection = "perspective"
+                    float verticalAperture = ${USDZExport.FilmGauge.toPrecision(USDZExport.Precision)}
+                }`;
+        }
+    }
+
+    private static async _BuildMaterials(materials: { [id: string]: Material }, textures: { [id: string]: Texture }, quickLookCompatible: boolean = false) {
+        const array = [];
+        for (const uuid in materials) {
+            const material = materials[uuid];
+
+            array.push(await USDZExport._BuildMaterial(material, textures, quickLookCompatible));
+        }
+
+        return `def "Materials"
+    {
+    ${array.join("")}
+    }
+    `;
+    }
+
+    private static _ChannelIndexMap: { [key: string]: number } = {
+        r: 0,
+        g: 1,
+        b: 2,
+        a: 3,
+    };
+
+    private static async _CreateDynamicTextureFromChannel(texture: Texture, channel: string, scene: Scene): Promise<DynamicTexture> {
+        const dt = new DynamicTexture("dt", { width: texture.getSize().width, height: texture.getSize().height }, scene, false);
+        const ctx = dt.getContext();
+        const img = (await texture.readPixels()) as any;
+        const data = (ctx as any).createImageData(texture.getSize().width, texture.getSize().height);
+        const count = texture.getSize().width * texture.getSize().height * 4;
+
+        for (let i = 0; i < count; i += 4) {
+            const value = channel === "r" ? 255 - img[i + USDZExport._ChannelIndexMap[channel]] : img[i + USDZExport._ChannelIndexMap[channel]];
+            data.data[i] = value;
+            data.data[i + 1] = value;
+            data.data[i + 2] = value;
+            data.data[i + 3] = 255;
+        }
+
+        ctx.putImageData(data, 0, 0);
+        dt.update(texture.invertY);
+        return dt;
+    }
+
+    private static async _BreakApartMetallicRoughnessAo(texture: Texture, scene: Scene): Promise<{ ao: DynamicTexture; roughness: DynamicTexture; metallic: DynamicTexture }> {
+        const ao = await USDZExport._CreateDynamicTextureFromChannel(texture, "r", scene);
+        ao.name = "ao";
+        const roughness = await USDZExport._CreateDynamicTextureFromChannel(texture, "g", scene);
+        roughness.name = "roughness";
+        const metallic = await USDZExport._CreateDynamicTextureFromChannel(texture, "b", scene);
+        metallic.name = "metallic";
+        return { ao, roughness, metallic };
+    }
+
+    private static async _BuildMaterial(_material: Material, textures: { [id: string]: Texture }, quickLookCompatible = false) {
+        // https://graphics.pixar.com/usd/docs/UsdPreviewSurface-Proposal.html
+        const pad = "			";
+        const inputs = [];
+        const samplers = [];
+        const material = _material as PBRMaterial;
+        const scene = material.getScene();
+
+        //TODO: rescope this whole function.

```

Comment:
What does that mean / imply?

Reply:
That the while function should be pulled out of the nested declaration but that would require a huge refactor which I am not doing.  I was asked to convert that three.js version which this was a major part of how they scoped this function.  I don't like it being nested but, am not going to tear it out this PR.

Reply:
I don't see any problem with that, and as you say it's how they did it in the Threejs code and we are simply doing a port, so I guess you can remove the comment.